### DISCUSSION TITLE
SLIP on NRF HW

### DISCUSSION THREAD

**Discussion by pbtgit** [zephyrproject-rtos/zephyr#48697] [Q&A]
I have successfully used SLIP with QEMU on a Linux host.
Now i want to run SLIP on NRF52840DK HW but it does not work.
The NRF board does not have Ethernet HW so the SLIP driver uses  NET_DEVICE_INIT with  DUMMY_L2 layer instead of ETH_NET_DEVICE_INIT. 

Assuming the DUMMY_L2 interace could have problems with Ethernet frames i changed the tunslip6 setup on linux side to use a tunnel device instead of the default tap device but still no go.

**Comment by jukkar** [comment]
I have not tested this kind of setup, but there should be no difference in zephyr side whether the slip is running in qemu or in target device. Note the dummy L2 in zephyr side does not add or use any L2 header so in linux side, a TUN device must be used (and not TAP).
Could you for example run a wireshark to your TUN device and see if there is any data transferred?

**Reply by pbtgit** [reply]
I agree that SLIP should not be dependent on QEMU or target. 
It turned out i was wrong, in my settings NET_L2_ETHERNET is actually defined so i should use TAP.
After wireshark sessin it seems like there is an issue with size or possibly bitrate:

- ARP seems to work, 
- Ping requests from Zephyr and gets response from host
- UDP with small payload from host works, larger payload does not
- Ping requests from host (size 0) gets response from zephyr

In the cases with lost (large) packets from host getting lost, i see no sign of them  being detected in slip driver in Zephyr.

**Reply by pbtgit** [reply]
I can observe that SLIP packets smaller than ~62 bytes works which explains why ARP and small ICMP requests work.
Lowering baudrate to /dev/ttyACMx in host does not help but adding per byte delay does work.
So i should look into HWFC to fix this.

**Reply by pbtgit** [reply]
I could not ghet this working with the USB CDC served by Nordic MCU.
However, when switching to the USB CDC served by Zephyr it actually works!

**Reply by krish2718** [reply]
@pbtgit I am stuck in the same situation as you, unable to get SLIP working, I had enabled USB CDC as well. Can you please elaborate on difference b/w Nordic USB CDC vs Zephyr USDB CDC?

**Reply by krish2718** [reply]
I have found the problem, SHELL wasn't completely disabled, so, `shell_thread` was stuck waiting for UART access, once I have completely disabled SHELL, ping worked straightaway even without any USB CDC, just plain SLIP + UART. Thanks.

**Reply by ofirshe** [reply]
Hey @krish2718 @pbtgit, could you share your configuration for setting up SLIP over UART? I’m currently facing an issue where pings sent from my Linux machine to the board don’t receive a response. The configuration I’m using can be found [here](https://github.com/zephyrproject-rtos/zephyr/discussions/82411#discussioncomment-11607813).

**Reply by krish2718** [reply]
Please see https://github.com/nrfconnect/sdk-nrf/tree/main/samples/wifi/wfa_qt_app and the command is
```
west build -b nrf7002dk/nrf5340/cpuapp -- -DEXTRA_CONF_FILE=overlay-slip.conf -DEXTRA_DTC_OVERLAY_FILE=nrf7002_uart_pipe.overlay
```

**Reply by ofirshe** [reply]
Thank you so much, @krish2718! I used your example, and it worked perfectly. I truly appreciate your help!

**Reply by AndreyDodonov-EH** [reply]
@ofirshe  @krish2718 Thanks a lot for all the information you provided!
But how did you end up commuinicating to the device?
Here https://github.com/zephyrproject-rtos/zephyr/discussions/82411 you @ofirshe  mention that you use `sudo slattach -v -s 115200 -p cslip -L /dev/ttyS2`
So what traffic does go through serial, IP or Ethernet?

Since there is no actual NET_L2 SLIP and we land in `ethernet_recv` and do things like NET_ETH_HDR, I'd expect that it should be Ethernet, but it is not how SLIP is supposed to work, directly encapsulating IP, or am I missing something?

Why is there no actual NET_L2_SLIP, by the way? Same as NET_L2_PPP exists?
@jukkar @rlubos would appreaciate your view here

**Reply by ofirshe** [reply]
@AndreyDodonov-EH 
On the Host side:
```
1. Install `tunslip6` by cloning and building the net-tools repository:
   git clone https://github.com/zephyrproject-rtos/net-tools.git
   cd net-tools/
   make

2. Create the SLIP interface using:
   ./tunslip6 -s <serial_port> -T <IPv6_prefix>

   For example:
   ./tunslip6 -s /dev/ttyS2 -T 2001:db8::1/64
```
On zephyr, I set the following SLIP related configurations:
```
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
CONFIG_NET_CONFIG_MY_IPV4_NETMASK="255.255.255.0"
CONFIG_NET_CONFIG_MY_IPV4_GW=""
CONFIG_NET_DRIVERS=y
CONFIG_SLIP=y
CONFIG_SLIP_TAP=y
CONFIG_SLIP_MAC_ADDR="00:00:00:00:00:01"
```

**Reply by AndreyDodonov-EH** [reply]
@ofirshe  Thanks for a quick response! 
I see you are using `-T` option here and `CONFIG_SLIP_TAP=y`, meaning that you are effectively using TAP, transferring Ethernet frames via SLIP. Do I understand correctly?
Do you know if it's possible for Zephyr to do TUN, i.e. transfer IP frames over SLIP? Will then DUMMY NET L2 be used?

**Reply by ofirshe** [reply]
Yes, I transmit Ethernet frames via SLIP. Regarding your second question, I have no answer.

**Reply by jukkar** [reply]
> Since there is no actual NET_L2 SLIP and we land in `ethernet_recv` and do things like NET_ETH_HDR, I'd expect that it should be Ethernet, but it is not how SLIP is supposed to work, directly encapsulating IP, or am I missing something?

If you want to transfer Ethernet frames, then the TAP interface should be created in the Linux side, and you should select `CONFIG_SLIP_TAP=y` to enable it (this is the default if SLIP is enabled). If you want to avoid L2 header, then `CONFIG_SLIP_TAP=n` should be set.

> 
> Why is there no actual NET_L2_SLIP, by the way? Same as NET_L2_PPP exists? @jukkar @rlubos would appreaciate your view here

There is no need for such layer as `NET_L2_DUMMY` exists already and it handles the case where we have no L2 (so when using TUN).

**Reply by jukkar** [reply]
> Do you know if it's possible for Zephyr to do TUN, i.e. transfer IP frames over SLIP? Will then DUMMY NET L2 be used?

It should be possible, you just need to setup the Linux side properly and disable TAP i.e., `CONFIG_SLIP_TAP=n`
